### PR TITLE
Integrating tromp solvers into miner

### PR DESCRIPTION
**Pull request by sarath-hotspot**
Integrated tromp equihash algorithm for mining. 
Original solvers are available in https://github.com/tromp/equihash

Forum thread: https://forum.z.cash/t/tromps-solvers/2465/197

Able to generate multiple blocks on testnet.
https://explorer.testnet.z.cash/block/000a8464db4d2e6ab597acdda0237641ac0baa4e4cd629a02fa47b06fa31a2079

By default this solver is disabled. Need to add macro declaration to enable this solver. 

How to enable tromp equihash solver:
In miner.cpp uncomment following line(35th line).
// #define USE_TROMP_EQUIHASH

Rebuild and run your zcshd.

Performance:
Single thread, on 4core, 8GB RAM machine,  getting 1 Sol/s.

### CODE REVIEW

**Comment by str4d** [comment]
Firstly, I've built this, it works (I mined several blocks), and it feels ridiculous (compared to my implementation). Thanks for integrating it @sarath-hotspot! :smile:

I discussed this with @nathan-at-least, and we're very keen to get this merged in before launch, as well as any other solvers that are available and can be integrated. One of our desired goals is for mining to be decentralised, and having several different solvers built into `zcashd` is a good way to reduce the barrier to users quickly getting started with mining (without needing to get into standalone miners, pools etc).

(And yet another disclaimer: merging of this PR, or of any solvers submitted to the Miner Challenge, into `zcashd` **does not** constitute an endorsement or have any correlations with eventual judging results by myself or the other judges. All Challange submissions are welcome to have PRs considered for merging.)

The primary issue now is getting this done in time for RC 2, which looks to be the last code merge point before launch. I'm going to review this PR now, but there is definitely some work that needs to be done before we merge it. @sarath-hotspot, will you have time to work on this PR over the next 30 or so hours? Comments will need to be addressed by the end of Thursday US if we are to merge this PR on Friday. If you don't have time, that's fine - let me know, and I can do the remaining integration work myself. But as the OP, it would be fantastic to have your contributions directly merged :relaxed:

File: src/Makefile.am [hunk 1]
```diff
@@ -224,7 +224,7 @@ libbitcoin_server_a_SOURCES = \
   validationinterface.cpp \
   $(JSON_H) \
   $(BITCOIN_CORE_H) \
-  $(LIBZCASH_H)
+  $(LIBZCASH_H)
```
Comment (str4d) [inline src/Makefile.am]: Remove the extra space.

File: src/Makefile.am [hunk 2]
```diff
@@ -243,6 +243,18 @@ libbitcoin_wallet_a_SOURCES = \
   $(BITCOIN_CORE_H) \
   $(LIBZCASH_H)
 
+TROMPEQUIHASH_SOURCES = \
```
Comment (str4d) [inline src/Makefile.am]: I'd prefer to use `src/equihash/SOLVERNAME` for imported solver code, so this code would be located in `src/equihash/tromp`.

File: src/miner.cpp [hunk 3]
```diff
@@ -30,6 +31,10 @@
 #include <boost/tuple/tuple.hpp>
 #include <mutex>
 
+#ifndef USE_TROMP_EQUIHASH
+// #define USE_TROMP_EQUIHASH
+#endif
```
Comment (str4d) [inline src/miner.cpp]: We expect that many users will just be using pre-built binaries (like our Debian packages), so instead we should use a runtime flag to determine which solver is run (see example below).

File: src/miner.cpp [hunk 4]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
                     std::lock_guard<std::mutex> lock{m_cs};
                     return cancelSolver;
                 };
+
+
+#ifdef USE_TROMP_EQUIHASH
```
Comment (str4d) [inline src/miner.cpp]:
```cpp
if (solver == "tromp") {
```

File: src/miner.cpp [hunk 5]
```diff
@@ -504,29 +509,6 @@ void static BitcoinMiner(CWallet *pwallet)
...
 
             while (true) {
-                // Hash state
-                crypto_generichash_blake2b_state state;
-                EhInitialiseState(n, k, state);
-
-                // I = the block header minus nonce and solution.
-                CEquihashInput I{*pblock};
-                CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
-                ss << I;
```
Comment (str4d) [inline src/miner.cpp]: Leave this here, since it is common code between all solvers.

File: src/miner.cpp [hunk 6]
```diff
@@ -504,29 +509,6 @@ void static BitcoinMiner(CWallet *pwallet)
...
-                // H(I||V||...
-                crypto_generichash_blake2b_state curr_state;
-                curr_state = state;
-                crypto_generichash_blake2b_update(&curr_state,
-                                                  pblock->nNonce.begin(),
-                                                  pblock->nNonce.size());
-
-                // (x_1, x_2, ...) = A(I, V, n, k)
-                LogPrint("pow", "Running Equihash solver with nNonce = %s\n",
-                         pblock->nNonce.ToString());
```
Comment (str4d) [inline src/miner.cpp]: Leave this here, and change to
```cpp
std::string solver = GetArg("-equihashsolver", "default");
LogPrint("pow", "Running Equihash solver \"%s\" with nNonce = %s\n",
         solver, pblock->nNonce.ToString());

File: src/miner.cpp [hunk 7]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
                 };
+
+
+#ifdef USE_TROMP_EQUIHASH
+            	//////////////////////////////////////////////////////////////////////////
+            	// TROMP EQ SOLVER START
+              	// I = the block header minus nonce and solution.
+                CEquihashInput I{*pblock};
+                CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+                ss << I;
```
Comment (str4d) [inline src/miner.cpp]: Remove the above (replaced with earlier left code).

File: src/miner.cpp [hunk 8]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+#else
+                // Use default mining algorithm.
+                // Hash state
+                crypto_generichash_blake2b_state state;
+                EhInitialiseState(n, k, state);
+
+                // I = the block header minus nonce and solution.
+                CEquihashInput I{*pblock};
+                CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+                ss << I;
```
Comment (str4d) [inline src/miner.cpp]: Remove the above (replaced with earlier left code).

File: src/Makefile.am [hunk 9]
```diff
@@ -261,8 +273,9 @@ crypto_libbitcoin_crypto_a_SOURCES = \
   crypto/sha256.cpp \
   crypto/sha256.h \
   crypto/sha512.cpp \
-  crypto/sha512.h
-
+  crypto/sha512.h \
+  ${TROMPEQUIHASH_SOURCES} 
+
```
Comment (str4d) [inline src/Makefile.am]: Remove extra spaces.

File: src/miner.cpp [hunk 10]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                // Nonce
+                // Nonce is u256 convert it to u32
+                const char *nonce = (const char*) pblock->nNonce.begin();
+
+              	// Create solver and initialize it with header and nonce.
+              	equi eq(1);
+              	eq.setnonce(header, headerLen, nonce, pblock->nNonce.size());
+
+                LogPrintf("Running TrompEquihash solver with nNonceInBlock=%s\n",
+                  		 pblock->nNonce.ToString());
```
Comment (str4d) [inline src/miner.cpp]: Remove this (replaced with earlier log line).

**Comment by sarath-hotspot** [comment]
Please let me know your comments. I could squeeze 3-4 hours from my work
schedule.
If I am done with the changes I will update the PR.

File: src/miner.cpp [hunk 11]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                eq.xfull = eq.bfull = eq.hfull = 0;
+                eq.showbsizes(0);
+                for (u32 r = 1; r < WK; r++) {
+                	r&1 ? eq.digitodd(r, 0) : eq.digiteven(r, 0);
+                	eq.xfull = eq.bfull = eq.hfull = 0;
+                	eq.showbsizes(r);
+                }
+                eq.digitK(0);
+
+                // Convert solution indices to charactar array(decompress) and pass it to validBlock method.
```
Comment (str4d) [inline src/miner.cpp]: "character array (decompress)"

File: src/miner.cpp [hunk 12]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                u32 nsols = 0;
+                for (unsigned s = 0; s < eq.nsols; s++) {
+                	nsols++;
```
Comment (str4d) [inline src/miner.cpp]: Replace tabs with spaces.

File: src/miner.cpp [hunk 13]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                eq.showbsizes(0);
+                for (u32 r = 1; r < WK; r++) {
+                	r&1 ? eq.digitodd(r, 0) : eq.digiteven(r, 0);
+                	eq.xfull = eq.bfull = eq.hfull = 0;
+                	eq.showbsizes(r);
+                }
+                eq.digitK(0);
+
+                // Convert solution indices to charactar array(decompress) and pass it to validBlock method.
+                u32 nsols = 0;
```
Comment (str4d) [inline src/miner.cpp]: `size_t`

File: src/miner.cpp [hunk 14]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                for (unsigned s = 0; s < eq.nsols; s++) {
+                	nsols++;
+                    LogPrintf("Checking solution %d \n", nsols);
+                    std::vector<eh_index> index_vector(PROOFSIZE);
+                    for (u32 i = 0; i < PROOFSIZE; i++) {
```
Comment (str4d) [inline src/miner.cpp]: `size_t`

File: src/miner.cpp [hunk 15]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                const char *nonce = (const char*) pblock->nNonce.begin();
+
+              	// Create solver and initialize it with header and nonce.
+              	equi eq(1);
+              	eq.setnonce(header, headerLen, nonce, pblock->nNonce.size());
+
+                LogPrintf("Running TrompEquihash solver with nNonceInBlock=%s\n",
+                  		 pblock->nNonce.ToString());
+
+                // Intialization done, start algo driver.
```
Comment (str4d) [inline src/miner.cpp]: This section looks like the actual Equihash algorithm. It should be in a method somewhere in `src/equihash/tromp`, ideally using the same API as the current solver or the [Miner Challenge API](https://zcashminers.org/rules).

File: src/miner.cpp [hunk 16]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                for (u32 r = 1; r < WK; r++) {
+                	r&1 ? eq.digitodd(r, 0) : eq.digiteven(r, 0);
+                	eq.xfull = eq.bfull = eq.hfull = 0;
+                	eq.showbsizes(r);
+                }
+                eq.digitK(0);
+
+                // Convert solution indices to charactar array(decompress) and pass it to validBlock method.
+                u32 nsols = 0;
+                for (unsigned s = 0; s < eq.nsols; s++) {
```
Comment (str4d) [inline src/miner.cpp]: `size_t`

File: src/miner.cpp [hunk 17]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                  		 pblock->nNonce.ToString());
+
+                // Intialization done, start algo driver.
+                eq.digit0(0);
+                eq.xfull = eq.bfull = eq.hfull = 0;
+                eq.showbsizes(0);
+                for (u32 r = 1; r < WK; r++) {
+                	r&1 ? eq.digitodd(r, 0) : eq.digiteven(r, 0);
+                	eq.xfull = eq.bfull = eq.hfull = 0;
+                	eq.showbsizes(r);
```
Comment (str4d) [inline src/miner.cpp]: I'd say that this needs at least one call to check for cancellation inside the `for` loop, but perhaps given the speed of the solver this isn't necessary (or at least, we should check what the effect is on the speed of the solver). We can probably get away with just checking for cancellation before starting the solver.

File: src/miner.cpp [hunk 18]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                    	// If we find a POW solution, do not try other solutions
+                      	// because they become invalid as we created a new block in blockchain.
+                      	  break;
+                    }
+                }
+                //////////////////////////////////////////////////////////////////////
+                // TROMP EQ SOLVER END
+                //////////////////////////////////////////////////////////////////////
+
+#else
```
Comment (str4d) [inline src/miner.cpp]:
```cpp
} else { // Default solver
```

File: src/miner.cpp [hunk 19]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                // H(I||V||...
+                crypto_generichash_blake2b_state curr_state;
+                curr_state = state;
+                crypto_generichash_blake2b_update(&curr_state,
+                                                  pblock->nNonce.begin(),
+                                                  pblock->nNonce.size());
+
+                // (x_1, x_2, ...) = A(I, V, n, k)
+                LogPrint("pow", "Running Equihash solver with nNonce = %s\n",
+                         pblock->nNonce.ToString());
```
Comment (str4d) [inline src/miner.cpp]: Remove this (replaced with earlier log line).

File: src/miner.cpp [hunk 20]
```diff
@@ -568,6 +630,7 @@ void static BitcoinMiner(CWallet *pwallet)
                     std::lock_guard<std::mutex> lock{m_cs};
                     cancelSolver = false;
                 }
+#endif
```
Comment (str4d) [inline src/miner.cpp]:
```
}
```

**Comment by str4d** [comment]
Note that I haven't reviewed the solver code itself. For launch, we would leave the default solver as the current well-reviewed code, but users could set a runtime flag to choose a different maybe-not-as-well-reviewed solver.

**Comment by daira** [comment]
Concept ACK; starting review now.

File: src/trompequihash/blake/blake2-config.h [hunk 21]
```diff
@@ -0,0 +1,72 @@
+/*
```
Comment (daira) [inline src/trompequihash/blake/blake2-config.h]: Given the changes @str4d described, I think there is no longer any need to include a BLAKE2b implementation separate from the one we use from libsodium, so strip out all of that code.

File: src/trompequihash/equi_miner.h [hunk 22]
```diff
@@ -0,0 +1,599 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include <assert.h>
+#include "util.h"
+
+typedef uint16_t u16;
+typedef uint64_t u64;
+
+#ifdef ATOMIC
```
Comment (daira) [inline src/trompequihash/equi_miner.h]: Prefix any preprocessor variables specific to this implementation with ``EQUIHASH_TROMP_``.

File: src/trompequihash/blake/blake2-config.h (hunk 21 above)
Comment (tromp) [inline src/trompequihash/blake/blake2-config.h]: The libsodium blake2b_update calls blake2b_compress only after it has at least 256 input bytes, preventing the extraction of a midstate.
My own copy of blake is modified to call blake2b_compress eagerly and reduces the buffer to 128 bytes.

**Comment by daira** [comment]
Note re: @str4d's comment https://github.com/zcash/zcash/pull/1570#pullrequestreview-4944785 : if the BLAKE2 code isn't present after refactoring then you don't need to add it in the first commit.

File: src/miner.cpp (hunk 17 above)
Comment (daira) [inline src/miner.cpp]: Note that if we move to more computationally intensive parameters, that could affect the decision of whether to check for cancellation per round or just once. For now, it's fine to just do whatever is simplest; it won't be hard to change.

File: src/miner.cpp (hunk 11 above)
Comment (daira) [inline src/miner.cpp]: "byte array (decompress)"

(Wide and multi-byte characters exist.)

**Comment by daira** [comment]
This looks excellent! I'll work on removing the BLAKE2 implementation now.

File: src/miner.cpp (hunk 6 above)
Comment (daira) [inline src/miner.cpp]: ``solver.c_str()``

(``c_str()`` returns a pointer to the string storage that does not need to be freed.)

File: src/miner.cpp [hunk 23]
```diff
@@ -559,6 +541,86 @@ void static BitcoinMiner(CWallet *pwallet)
...
+                	eq.xfull = eq.bfull = eq.hfull = 0;
+                	eq.showbsizes(r);
+                }
+                eq.digitK(0);
+
+                // Convert solution indices to charactar array(decompress) and pass it to validBlock method.
+                u32 nsols = 0;
+                for (unsigned s = 0; s < eq.nsols; s++) {
+                	nsols++;
+                    LogPrintf("Checking solution %d \n", nsols);
```
Comment (daira) [inline src/miner.cpp]: ``LogPrint("pow", "Checking solution %d\n", nsols);``

**Comment by sarath-hotspot** [comment]
@daira  I see you already submitted another PR with updated changes.  I will skip making these changes. 
https://github.com/zcash/zcash/pull/1578

Thanks for taking care of changes. 

Are you planning to modify libsodium source code?

**Comment by tromp** [comment]
> @daira I see you already submitted another PR with updated changes. I will skip making these changes.
> #1578
>
> Thanks for taking care of changes.
>
> Are you planning to modify libsodium source code?

The impact of the blake change may be less than that of the speed
improvements I made after you snapshotted my code. Please consider
grabbing the latest code. Then again, I expect to have more speedups
today...

-John

**Comment by str4d** [comment]
@tromp #1578 uses the latest code from your repo.

**Comment by str4d** [comment]
Superceded by #1578.

**Pull request closed without merging**